aspnetcore/grpc/services.md

```diff
@@ -268,6 +268,68 @@ public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCa
 }
 ```
 
+## Multi-threading with gRPC streaming methods
+
+There are important considerations to gRPC streaming methods that use multiple threads.
+
+### Reader and writer thread safety
+
+`IAsyncStreamReader<TMessage>` and `IServerStreamWriter<TMessage>` can each be used by only one thread at a time. For a streaming gRPC method, multiple threads can't read new messages with `requestStream.MoveNext()` simultaneously. And multiple threads can't write new messages with `responseStream.WriteAsync(message)` simultaneously.
+
+A safe way to enable multiple threads to interact with a gRPC method is to use the producer-consumer pattern with [System.Threading.Channels](/dotnet/core/extensions/channels).
+
+```csharp
+public override async Task DownloadResults(DataRequest request,
+    IServerStreamWriter<DataResult> responseStream, ServerCallContext context)
+{
+    var channel = Channel.CreateBounded<DataResult>(new BoundedChannelOptions(capacity: 5));
+
+    var processTask = Task.Run(async () =>
+    {
+        var dataChunks = request.Value.Chunk(size: 10);
+
+        // Process results in multiple background tasks.
+        var processTasks = dataChunks.Select(
+            async c =>
+            {
+                // Write results to channel across different threads.
+                var message = new DataResult { BytesProcessed = c.Length };
+                await channel.Writer.WriteAsync(message);
+            });
+        await Task.WhenAll(processTasks);
+
+        channel.Writer.Complete();
+    });
+
+    // Read results from channel.
+    await foreach (var message in channel.Reader.ReadAllAsync())
+    {
+        await responseStream.WriteAsync(message);
+    }
+
+    await processTask;
```

Comment:
I know it's in a comment below, but maybe a code comment would be good to about making sure to wait for everything to complete before exiting the method.